src/polydatum/dal.py

```diff
@@ -1,20 +1,55 @@
 from __future__ import absolute_import
+
+import inspect
 from contextlib import contextmanager
-from .context import DataAccessContext
-from polydatum.errors import AlreadyExistsException
+
+from polydatum.middleware import PathSegment, DalCommandRequest, DalCommand, dal_resolver, \
+    dal_method_resolver_middleware, handle_dal_method
+from polydatum.errors import AlreadyExistsException, InvalidMiddleware
 from polydatum.util import is_generator
 from .resources import ResourceManager
 from .context import _ctx_stack
+from typing import Callable, Tuple
+from functools import partial, update_wrapper
+from polydatum.context import DataAccessContext
 
 
 class DataAccessLayer(object):
     """
     Gives you access to a DataManager's services.
     """
 
-    def __init__(self, data_manager):
+    # default middleware classes need to be instantiated before being
+    # passed to the init function because the resulting object is called
+    # directly (__call__).
+    def __init__(
+        self,
+        data_manager,
+        middleware=None,
+        default_middleware=(dal_method_resolver_middleware,),
+        handler=handle_dal_method
+    ):
         self._services = {}
         self._data_manager = data_manager
+        self._handler = handler
+        reversed_middleware = []
+
+        middleware = middleware or []
+        if default_middleware:
+            middleware.extend(default_middleware)
+        for m in reversed(middleware):
+            if inspect.isclass(m):
+                m = m()
+            if not isinstance(m, Callable):
+                raise InvalidMiddleware(f'{m} is not a valid Callable middleware')
+            reversed_middleware.append(m)
```

Comment:
nitpick: `reversed_middleware` not really needed. You could just use `reversed(middleware)` again. Or at the start do: `reversed_middleware = reversed(middleware)` instead of building a list in a loop.